notebooks/02_model/vowpal_wabbit_deep_dive.ipynb

```diff
@@ -0,0 +1,737 @@
+      "RSQUARED: 0.09877954418265689\n",
+      "EXP_VAR: 0.12455611599414951\n",
+      "CPU times: user 2.11 s, sys: 31.7 ms, total: 2.14 s\n",
+      "Wall time: 4.61 s\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%time \n",
+    "\n",
+    "train_params = 'vw --loss_function logistic --oaa 5 -f {model} -d {data}'.format(model=model_path, data=train_path)\n",
+    "test_params = 'vw -i {model} -d {data} -t -p {pred}'.format(model=model_path, data=test_path, pred=prediction_path)\n",
+    "\n",
+    "run_vw(train_params=train_params,\n",
+    "       test_params=test_params,\n",
+    "       test_data=test,\n",
+    "       prediction_path=prediction_path)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Additionally, one might simply be interested in whether the user likes or dislikes an item and we can adjust the input data to represent a binary outcome, where ratings in (1,3] are dislikes (negative results) and (3,5] are likes (positive results).\n",
+    "\n",
+    "This framing allows for a simple logistic regression model to be applied. To perform logistic regression the loss_function parameter is changed to 'logistic' and the target label is switched to [0, 1]."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "RMSE: 1.764643840355633\n",
+      "MAE: 1.4427961357602124\n",
+      "RSQUARED: -1.490189004814928\n",
+      "EXP_VAR: -0.07611382892527008\n",
+      "CPU times: user 2.05 s, sys: 43.4 ms, total: 2.09 s\n",
+      "Wall time: 2.9 s\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%time \n",
+    "\n",
+    "train_params = 'vw --loss_function logistic -f {model} -d {data}'.format(model=model_path, data=train_logistic_path)\n",
+    "test_params = 'vw -i {model} -d {data} -t -p {pred}'.format(model=model_path, data=test_logistic_path, pred=prediction_path)\n",
+    "\n",
+    "run_vw(train_params=train_params,\n",
+    "       test_params=test_params,\n",
+    "       test_data=test,\n",
+    "       prediction_path=prediction_path)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "So far we have treated the user features and item features independently, but taking into account interactions between features can provide a mechanism to learn more fine grained preferences of the users.\n",
+    "\n",
+    "To generate interaction features use the quadratic command line argument and specify the namespaces that should be combined: '-q ui' combines the user and item namespaces based on the first letter of each.\n",
+    "\n",
+    "When generating interaction terms one thing to consider is the [hash space](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Feature-Hashing-and-Extraction) used for the features. It can be beneficial to increase the size of the space to reduce unwanted collisions."
```

Comment:
I think it would be good to explain briefly what is the hashing mechanism in VW and what are its benefits